[feature]: Metadata package comparator/merger

JorgeMMdev opened the pull request:
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/8698w42jn

### :memo: Implementation
- [x] JSON Comparator/Merge tool: Compares two JSON files, shows the differences and gives users options to select what changes go into the merged JSON file.


### :art: Screenshots



https://github.com/user-attachments/assets/057a3472-862e-4a48-9a0e-8f054d344863



### :fire: Notes to the tester
Start comparator with:
```
yarn compare-metadata -f=`path-to-first-file` -s=`path-to-second-file`
```
where `-u` and `-s` are both optional parameters. Alternatively, we can start the comparator with the JSON path parameters using:
```
yarn compare-metadata -first=`path-to-first-file` -second=`path-to-second-file`
```

Files to test with:
[test-comparator1.json](https://github.com/user-attachments/files/23631548/test-comparator1.json)
[test-comparator2.json](https://github.com/user-attachments/files/23631551/test-comparator2.json)


#8698w42jn

JorgeMMdev committed:
React App config

JorgeMMdev committed:
Resolve error of the app config

JorgeMMdev committed:
fix app launch

JorgeMMdev committed:
jsondiffpatch lib test

JorgeMMdev committed:
Clean Dhis2 dependencies

JorgeMMdev committed:
remove d2-api dependencies

JorgeMMdev committed:
clean unused components

JorgeMMdev committed:
Basic componentes + upload containers

JorgeMMdev committed:
json containers and improved app structure

JorgeMMdev committed:
improved app structure

deeonwuli committed:
feat: add json difference to comparator box

deeonwuli committed:
feat: update title in index.html and clean up unused code

deeonwuli committed:
feat: start comparator only if jsons are not equal

deeonwuli committed:
feat: refactor JSON comparison logic and implement file download functionality

deeonwuli committed:
feat: enhance JSON comparison functionality with new components and utilities

deeonwuli committed:
feat: add useDiffLineProps hook and refactor DiffLine component for improved state management

deeonwuli committed:
feat: reorganize hooks structure by moving useComparator, useDiffLineProps, and useDownloadJSON to hooks directory

deeonwuli committed:
fix: add error handling for JSON parsing in parseMetadataFromFile function

deeonwuli committed:
refactor: replace hardcoded colors with a colors object in useDiffLineProps hook

deeonwuli committed:
fix: add error handling for JSON parsing in parseJSONFromEnvInput function

deeonwuli committed:
refactor: simplify renderValue function

deeonwuli committed:
chore: add unit tests for renderValue function/fix failing test

deeonwuli committed:
fix: update command argument names for clarity in compare-metadata function

deeonwuli committed:
fix: improve error logging for JSON parsing

deeonwuli committed:
refactor: replace string literals with Choice enum

deeonwuli committed:
fix: replace JSON.parse with structuredClone for deep cloning and improve handling of undefined values in renderValue function

deeonwuli committed:
fix: update width property in FileInputWrapper and ensure handleFileChange dependency is set correctly

deeonwuli committed:
chore: update translation files

deeonwuli committed:
fix: remove unnecessary css property

ifoche reviewed (changes requested):
following our last conversation, @deeonwuli , I'm submitting a request for changes here, so it represents the changes we have discussed:
- Having an unsorted version of the 2 sorted files we want to merge
- Having a center panel representing the merged file, which starts by being a copy of the left side and then evolves with the users decisions
- The changes that are decided on the sorted files are applied to the unsorted ones
- We discussed to block the comparison until the 4 files are provided (unsorted and sorted versions) but I'm thinking that better than that, only the unsorted versions are to be provided, and the sorted version is calculated on-the-fly by the tool
- Every time an object is modified, that same object is located in the unsorted merged file and modified accordingly. 
- If a new object is added from one of the sorted files, the file is located in the unsorted version and placed there in the unsorted merged file.
- The final unsorted merged file is the one that should be downloadable, in order to be installed

deeonwuli committed:
feat: implement temporary file handling and refactor JSON fetching logic

deeonwuli committed:
chore: ignore .DS_Store and remove from repo

deeonwuli committed:
fix: update translation strings in MergedPreviewContainer

deeonwuli committed:
fix: update paths in build file to use process.cwd()

deeonwuli committed:
feat: allow different names for output files

derive output filename from the input folder name

deeonwuli committed:
refactor: change JSONContent from class to interface and update related methods

deeonwuli committed:
fix: update path resolution in build command to use path.resolve

deeonwuli committed:
feat: enable comparator only mode in Vite configuration to skip DHIS2 proxy and auth checks

deeonwuli committed:
Refactor JSON Comparator/Merge Component

- Removed DiffLine and DiffViewer components to streamline the codebase.
- Introduced DiffSection component to handle the diffing and merging logic.
- Updated useComparator hook to manage left and right JSON states and merging functionality.
- Implemented useJsonDiffSelector hook for managing JSON differences and selections.
- Enhanced ImportFile component to support optional labels and improved styling.
- Refined LandingPage layout and styling for better user experience.
- Adjusted theme provider integration in App component for consistent theming.

deeonwuli committed:
fix: update logo source in LandingPage component

deeonwuli committed:
refactor: convert function declarations to arrow functions in JSONContent and improve type handling in useJsonDiffSelector
feat: add json utility functions for sorting, formatting, parsing, and cloning JSON data
fix: update logo import in LandingPage and enhance layout structure

deeonwuli committed:
refactor: replace fast-json-patch with custom JSON comparison logic and remove unused ComparatorViewModel component

deeonwuli committed:
chore: install packages, update translation files

deeonwuli committed:
add empty capture directory

deeonwuli committed:
fix: sort array of objects by id first

deeonwuli committed:
fix: setNestedValue function to handle creating nested structures when left path is N/A

deeonwuli committed:
feat: implement restoreOriginalOrder function and update useComparator to maintain original order of merged JSON

nshandra commented on capture/DS_Store:
macOS system file, should be deleted AFAIK.

nshandra commented on package.json:
```diff
@@ -1,29 +1,138 @@
 {
     "name": "metadata-skeleton-offline-2",
```
Why metadata-skeleton-offline-2?
Its referenced in the README too, so i guess its a leftover from development.

deeonwuli replied:
You're right. Looks like a leftover from development

deeonwuli committed:
fix: ts terminal error

deeonwuli committed:
docs: update README with metadata comparator usage instructions and formatting improvements

nshandra commented on src/webapp/pages/app/__tests__/App.spec.tsx:
```diff
@@ -0,0 +1,16 @@
+    it("renders the feedback component", async () => {
+        const view = getView();
+
+        expect(await view.findByText("Send feedback")).toBeInTheDocument();
```
This test fails with: `TestingLibraryElementError: Unable to find an element with the text: Send feedback.`

nshandra commented on package.json:
```diff
@@ -1,29 +1,138 @@
+        "source-map-explorer": "yarn build --sourcemap=true && source-map-explorer --no-border-checks --only-mapped build/assets/*js",
+        "test": "yarn test-unit",
+        "test-fixtures-generate": "yarn run-script src/tests/manage.ts generate-fixtures",
+        "test-snapshots-delete": "yarn run-script src/tests/manage.ts delete",
```
These commands fail with 'MODULE_NOT_FOUND'.  Looks like a leftover.

nshandra commented on src/domain/usecases/VisualizationCombineAndRemoveDuplicatesUseCase.ts:
```diff
@@ -1,4 +1,4 @@
-import { ProcessedDataSetProgram } from "../entities/ProcessedDataSetProgram";
+
```
Styling: remove leading blank line.

nshandra commented on src/webapp/pages/app/__tests__/__snapshots__/App.spec.tsx.snap:
This snapshot is outdated, and no longer needed AFAIK.

nshandra commented on src/scripts/compare-metadata.ts:
```diff
@@ -0,0 +1,121 @@
+        },
+    });
+
+    run(cmd, process.argv.slice(2));
```
Question/nitpick: why  not use `cli.ts` ?
That will make it more in line with the rest of tools like the ones in d2-tools.

deeonwuli replied:
I overlooked `cli.ts`. Thanks for catching that. I've refactored to use `cli.ts`

nshandra commented on README.md:
```diff
@@ -4,7 +4,7 @@ This project processes metadata for datasets, programs, permissions, and visuali
 ## Prerequisites
 
-- Node.js (>=14.x)
+-   Node.js (>=14.x)
```
Does not match `.nvmrc`

nshandra reviewed (approved):
Overall, pretty good.
The comments are not blocking issues, some leftovers and incorrect node version in README, but the broken test should be fixed IMO.

deeonwuli committed:
Add capture folder structure

deeonwuli committed:
fix: update project name and repository links in README and package.json

deeonwuli committed:
test: remove obsolete App test files

deeonwuli committed:
refactor: remove obsolete test scripts from package.json and clean up VisualizationCombineAndRemoveDuplicatesUseCase

deeonwuli committed:
feat: implement metadata comparison command in CLI

nshandra commented on package.json:
```diff
@@ -1,29 +1,136 @@
+        "@vitejs/plugin-react": "^3.1.0",
+        "@vitest/coverage-v8": "^0.32.4",
+        "bundle-phobia-cli": "^0.14.14",
+        "cmd-ts": "^0.12.1",
```
Sorry i missed this in the previous review, but `cmd-ts` is repeated with diferent versions. Remove it from the dev section.

nshandra commented on package.json:
```diff
@@ -1,29 +1,136 @@
+        "react": "18.3.1",
+        "react-dom": "18.3.1",
+        "react-router": "5.2.1",
+        "react-router-dom": "6",
```
Major version mismatch.

nshandra commented on package.json:
```diff
@@ -1,29 +1,136 @@
+        "react-router-dom": "6",
+        "react-split": "^2.0.14",
+        "real-cancellable-promise": "^1.1.2",
+        "styled-components": "6.1.11",
```
Major version mismatch with `"@types/styled-components": "5.1.24"`
Also, looks like the `@types` its not needed with v6: [see](https://styled-components.com/docs/faqs)

deeonwuli replied:
Right. 👍  `@types/react-router` and `@types/react-router-dom` are also not needed in react-router v6: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/57705

nshandra reviewed (changes requested):
Previous comments resolved. I noticed some issues with the dependencies. 
Eventually a deeper analysis should be done to remove unused packages and fix the warnings reported by `yarn install`

deeonwuli committed:
refactor: update dependencies and remove obsolete packages

nshandra reviewed (approved):
All good, thanks @deeonwuli

xurxodev reviewed (changes requested):
thanks @deeonwuli

This PR adds the **metadata comparator app** (JSON diff with Monaco, manual merge, and download) and modifies the existing **metadata build** (combining capture, visualizations, and permissions into output/).

In general terms there are a good Clean Architecture implementation

## 1. Should fix

-   **`useComparator.tsx` – `useCallback` dependencies**  
    `uploadLeft`, `uploadRight`, `acceptLeft`, `acceptRight`, and `handleMergedChange` use `restoreOriginalMergedOrder` but do not include it in the dependency array. Add `restoreOriginalMergedOrder` (and any deps that callback needs) to avoid stale closures.

-   **`Share.tsx` – style typo**  
    Line 97: `borderradius: 0` → `borderRadius: 0` (camelCase and correct property name).

-   **`compare-metadata.ts` – redundant branch**  
    Lines 41–50: both the `if` and `else` end up calling `startMetadataComparator`; only the “identical” case does an early return. Simplify to: if both JSONs are valid and identical → log and return; otherwise → a single call to `startMetadataComparator(json1, json2)`.

-   **README – two functionalities not clearly separated**  
    The README does not clearly state that the project has **two distinct functionalities** (build vs compare).

-   **Script vs app – parse JSON**: The “string → Maybe<JSONContent>” logic is duplicated in the script (`parseMetadataFromFile`) and in the app (`parseJson` in jsonUtils). Consider extracting `parseJson` into a shared module (utils or domain) and using it in both; in `useJsonDiffSelector` replace direct `JSON.parse` with that function.
-   **`Share.tsx`**: Class component with inline styles; consider migrating to a function component + styled-components for consistency.

-   **pre-push hook**: EyeSeeTea projects use a **pre-push** hook (e.g. to run lint/test before push). This repo currently has no `pre-push` in `.husky/`. Add a `.husky/pre-push` that runs the same checks (e.g. `yarn lint && yarn prettify && yarn test` or `yarn check-code`) so it aligns with other EyeSeeTea repos.
-   **GitHub Action**: There is no `.github/workflows/` configuration reusing the shared workflow **`EyeSeeTea/github-workflows/.github/workflows/master.yml@master`** if required
-   **Remove Travis CI**: Only GitHub Actions should be used as CI. This repo has a `.travis.yml` file; remove it and rely solely on the GitHub Actions workflow above. Do not keep Travis (or any other CI system)

-   **Unify metadata entities with a common type (`MetadataItem`)**  
    Entities such as Category, DataElement, CategoryOptionCombo, LegendSet, Visualization, etc. are only used as the types of arrays in `ProcessedDataSetProgram` and `ProcessedVisualization`. The code does not access any of their properties beyond `id` (used in `concatUnique` for deduplication).  
    **Proposal:** In `common.ts`, define a type `MetadataItem = BaseMetadata & Record<string, unknown>`, type all those arrays in `ProcessedDataSetProgram` and `ProcessedVisualization` as `MetadataItem[]`, and remove the entity files that are left with no references (Category, DataElement, CategoryOption, CategoryOptionCombo, LegendSet, CategoryCombos, Section, DataSet, OptionSet, Option, Program, ProgramStage, ProgramRule, ProgramRuleActions, ProgramRuleVariable, ProgramIndicators, Dashboards, Indicators, LegenSetsVisuals, Visualizations, IndicatorType). Keep Ref, BaseMetadata, User, Users, UserGroups (for ProcessedPermissions and tests). Runtime JSON shape stays the same; only the type model is simplified and maintenance is reduced.

- **Use Future** - The repository and use cases is using Promise as return type instead of Future

## 2. Recommendations non blocking

-   Unit tests for `detectJsonDifferences` and for `restoreOriginalOrder` / `sortJSONKeys`.
-   Fix test copy in `User.spec.ts` (line 15): “true when the id exists” instead of “false when the id exist”.
-   prioritise domain tests for `jsonUtils` and `detectJsonDifferences`; the rest according to team strategy.

deeonwuli committed:
fix: improve ESLint configuration and fix dependency handling in useComparator hook

deeonwuli committed:
refactor: convert Share component from class to functional component

deeonwuli committed:
refactor: centralize JSON parsing logic by introducing parseJson utility

deeonwuli committed:
feat: add pre-push hook to automate code formatting, linting, and testing

deeonwuli committed:
feat: add GitHub workflows and pull request template, remove Travis CI configuration

deeonwuli committed:
feat: refactor metadata processing to use Future for async operations

deeonwuli committed:
feat: remove deprecated entity interfaces and replace with MetadataItem type

deeonwuli committed:
feat: update Readme, update compare-metadata command

deeonwuli committed:
fix: update JSON handling to use JSONValue type

ifoche reviewed (changes requested):
Hi @deeonwuli . 

This is very useful. I am, though, missing one thing yet:
While you merge, it looks like the unsorted merged file has some glitches. If you look at this screenshot, you'll see that even when I merge the options of the dataset that are in both files in the same order: Opt A and then Opt B, I add all options and it become Opt B first and Opt A last
<img width="1834" height="789" alt="Image" src="https://github.com/user-attachments/assets/7b16d611-0668-40ec-bc89-86f74c8ac3d6" />

deeonwuli committed:
fix: compare and merge json correctly for array of objects

deeonwuli committed:
tests: add tests for JSON utilities

ifoche reviewed (approved):
Sorry @deeonwuli . My bad. The order of options in the test1 raw file was actually like the one in the merge file. So the code is perfectly fine and correct to me from a functional perspective

xurxodev reviewed (approved):
thanks @deeonwuli 

Only a comment I don't see the check in the last commits indicating the github action was executed.

Maybe because the base branch is not development I don't know.

Talk with your PM to create a separate task to review in the furute

deeonwuli committed:
fix: update compare-metadata command to correct argument order

adrianq committed:
minor update in the README

adrianq commented:
Thanks a lot @deeonwuli @ifoche @nshandra @JorgeMMdev @xurxodev Finally merging this!
